### ISSUE TITLE
case splitting only works when hole is the last thing on the line

### ISSUE THREAD

**Issue by Snicksie** [agda/agda#2628]
When case splitting on `x` in the first hole in the function `add` below, additional holes get generated.

```agda
data ℕ : Set where
  zero : ℕ
  suc : ℕ → ℕ

nat = ℕ

data ℚ : Set where
  quot : ℕ → ℕ → ℚ

add : ℚ → ℚ → ℚ
add x y = quot {!!} {!!}
```

Result:

```
add : ℚ → ℚ → ℚ
add (quot x x₁) y = quot ? ? {!!}
```
I am using Agda version 2.5.2-16, as found in the Arch repositories.

**Comment by jespercockx** [comment]
I can confirm this bug still exists in the current stable branch.

**Comment by jespercockx** [comment]
I guess this was introduced by @andreasabel in the fix to #2051.

**Comment by andreasabel** [comment]
Yes, before my work on #2051 and friends you could only split when the rhs was a single ?.  In that sense, even the current buggy behavior is better than what we had before.
Splitting in the second hole works fine.

The following also does not work nicely. Splitting on `x` here
```
add : ℚ → ℚ → ℚ
add x y = quot {!x!} zero
```
produces rhs `quot ? zero zero`.

It seems that splitting works properly only when the ? is the last thing on the line.  Another example:
```agda
add : ℚ → ℚ → ℚ
add x y = quot {!!} (suc {!x!})
```
Here you get `quot ? (suc ?))` with an extra parenthesis.
It seems the emacs mode deletes from the beginning of the line until the question mark and replaces it with the produced clause.

Try for instance:
```agda
bla : ℚ → Set
bla x =
  {!x!}
```
You get
```agda
bla : ℚ → Set
bla x =
  bla (quot a b) = ?
```
I think this never worked.

Good luck with the milestone 2.5.3! :D

**Comment by andreasabel** [comment]
I think with the current infrastructure the emacs mode cannot do much better.  After the load the user is allowed to shift holes around and freely add Agda text and comments.  Agda does only knows the state of the last load action, but contents of the emacs buffer may have diverged from that when the user calls the case split tactics.  Agda sends the new clauses back and the emacs-mode has to integrate them somehow into the text which is different from what Agda perceived it was.

For this to work, the emacs-mode would have to keep track of the boundaries of a clause to be able to substitute it completely.  Agda could, after parsing, communicate to emacs where the clause boundaries are, and emacs puts some invisible markers there which move along with the text the user adds.

However, even then case split is not safe.  It could overwrite stuff the user put inside the clause boundaries.

I'd speculate the current buggy behavior predates Agda 2 and existed even in Agda 1.  As far as I know, the interaction model of Agda 2 was inherited from Agda 1.

**Comment by jespercockx** [comment]
Oh, I was kind of hoping there would be an obvious solution when I added the milestone, or at least something that would help in this case. But maybe it's not as easy as I thought...

I think case splitting should produce valid Agda code at least in those cases where there haven't been any changes since the last load action. For that case, we can just delete the whole right-hand side (even if it's multiple lines) and replace it by the thing generated by Agda, no?

**Comment by andreasabel** [comment]
Yes.  This could be tracked in the emacs mode and passed as a flag to the case tactic.
Also Agda then has to inform the emacs mode which part to replace.

**Comment by UlfNorell** [comment]
Fail when hole is on a different line is #1146.

**Comment by UlfNorell** [comment]
> I'd speculate the current buggy behavior predates Agda 2 and existed even in Agda 1. As far as I know, the interaction model of Agda 2 was inherited from Agda 1.

Historical note: Agda 1 did not have pattern matching left-hand sides, only `case` expressions.
